build/crd-samples/sedna/objectsearchservice_v1alpha1.yaml

```diff
@@ -0,0 +1,60 @@
+apiVersion: sedna.io/v1alpha1
+kind: ObjectSearchService
+metadata:
+  name: pedestrian-search
+  namespace: default
+spec:
+  userWorker:
```

Comment:
is there any parameters in user worker? what key-words should be exposed to others?

Reply:
I had a few ideas regarding additional parameters that could be useful in the near future. For example:

1. There could be a parameter expressing a **key requirement** for the end-user. For example, `{"requirement": "optimize_latency"}` could be interpreted by Sedna as the need for the service to run as quickly as possible rather than optimizing for other metrics such as accuracy. In fact, this could make the framework load a shallow model rather than a more complicated one (in fact favouring latency over precision).
2. For the models, the field in the YAML could be already an array rather than a single value to accommodate what is expressed in point 1.
3. A parameter called _mesh_enabled_ (True/False) could be used for each edge node to express the need to setup edge mesh connectivity (KubeEdge edgemesh lib).
4. Maybe a _priority_ parameter in case, one day, there could be multiple pipelines/pods running in parallel and it becomes crucial to allocate resources to a service rather than another based on its mission criticality.

These are just some ideas that I had in mind. Also, as a general suggestion, it would be useful to _compress_ a bit the YAML to reduce its verbosity. For instance, when multiple edge nodes share the same configuration, we should be able to write such configuration only once for all of them and pass the edge "cluster" using it as an array.

Finally, it would be very useful for Sedna to understand the capabilities of the available edge nodes in the cluster to optimally schedule tasks on them based on their resources. Right now, the pod allocation is **explicit** which means that we manually select on which node each container runs.
In the spirit of promoting deployment automation and make the framework more flexible, we could have some sort of **implicit** allocation. In this way, Sedna could actively decide where to schedule each pod and meet key requirement/s provided by the service issuer. This probably requires a longer discussion, but I wanted to share it anyway.

Reply:
> Right now, the pod allocation is explicit which means that we manually select on which node each container runs.

I think that we can just use the k8s schedule capability i.e. nodeSelector/nodeAffinity.

Reply:
> 3. A parameter called mesh_enabled (True/False) could be used for each edge node to express the need to setup edge mesh connectivity (KubeEdge edgemesh lib).

From https://github.com/kubeedge/edgemesh#test-case, it seems no need extra flag for edgemesh

Reply:
About the YAML compression, I'm posting below how I thought we could streamline the `trackingWorkers` section:

```
trackingWorkers:
    metadata:
        name: [trackingWorker0, trackingWorker1]
        spec:
            nodeSelector:
                edgeai/app: [worker0, worker1]
            containers: 
                - name: [trackingWoker0, trackingWoker1]
                  image: kubeedge/sedna-example-multi-edge-inference-reid-dataLoad-and-tracking:v0.1.0
            env:
                - name: "dataSources"
                  value: {
                      "trackingWoker0":[192.168.0.12,192.168.0.13],
                      "trackingWoker1":[192.168.0.10,192.168.0.11]
                      
                  }
                - name: "modelUri"
                  value: "http://103.124.124.134/tracking_v1.0.tar.gz"

```

Alternatively, this is a beautified version:

```
trackingWorkers:
  metadata:
    name:
      - trackingWorker0
      - trackingWorker1
    spec:
      nodeSelector:
        edgeai/app:
          - worker0
          - worker1
      containers:
        - name:
            - trackingWoker0
            - trackingWoker1
          image: >-
            kubeedge/sedna-example-multi-edge-inference-reid-dataLoad-and-tracking:v0.1.0
      env:
        - name: dataSources
          value:
            trackingWoker0:
              - 192.168.0.12
              - 192.168.0.13
            trackingWoker1:
              - 192.168.0.10
              - 192.168.0.11
        - name: modelUri
          value: 'http://103.124.124.134/tracking_v1.0.tar.gz'
```

For this to work, additional code would be probably needed in the pod creation function in the GM (the relative service file) in order to correctly parse the new YAML structure. This might be useful or not, depending on how many workers with the same specs we are spawning.

Reply:
We may want to reuse and expose the k8s pod template to the user, and your proposals modify the pod template. So I think a better way is that advancing the common field independent with pod template

Reply:
look like this:
```yaml
trackingWorkers:
  envs:
    - name: dataSources
      value:
      - 192.168.0.12
      - 192.168.0.13
    - name: dataSources
      value:
      - 192.168.0.10
      - 192.168.0.11      
  template:
    metadata:
    spec:
      containers:
      - name: 
        image:  kubeedge/sedna-example-multi-edge-inference-reid-dataLoad-and-tracking:v0.1.0
        env:
        - name: modelUri
          value: 'http://103.124.124.134/tracking_v1.0.tar.gz'
```

Reply:
How about this:
1) When we want to specify the physical node that the workers need to deploy and its corresponding data source.
```yaml
  trackingWorkers:
    dataSpec: 
      selectData:
        - nodeSelector:
            edgeai/app: worker0
          dataSources: [192.168.0.10, 192.168.0.11]
        - nodeSelector:
            edgeai/app: worker1
          dataSources: [192.168.0.12, 192.168.0.13]
    template:
      spec:
        containers: 
          - name: trackingWoker0
            image: kubeedge/sedna-example-multi-edge-inference-reid-dataLoad-and-tracking:v0.1.0
            env:
              - name: "modelUri"
                value: "http://103.124.124.134/tracking_v1.0.tar.gz"
```
2) When we we don't want to specify the physical node that the workers need to deploy and its corresponding data source.
```yaml
  trackingWorkers:
    dataSpec: 
      data:
        dataSources: [192.168.0.10, 192.168.0.11, 192.168.0.12, 192.168.0.13]
    template:
      spec:
        containers: 
          - name: trackingWoker0
            image: kubeedge/sedna-example-multi-edge-inference-reid-dataLoad-and-tracking:v0.1.0
            env:
              - name: "modelUri"
                value: "http://103.124.124.134/tracking_v1.0.tar.gz"
```

The user can choose one of the "data" and "selectData" in "dataSpec" to fill in.

Reply:
I agree with both comments!

Reply:
```yaml
  trackingWorkers:
    dataSpec: 
      selectData:
        - nodeSelector:
            edgeai/app: worker0
          dataSources: [192.168.0.10, 192.168.0.11]
        - nodeSelector:
            edgeai/app: worker1
          dataSources: [192.168.0.12, 192.168.0.13]
```

nodeSelector means trackingWoker placement nodes, so it should not be in dataSpec section.

```yaml
  trackingWorkers:
    dataSpec: 
      data:
        dataSources: [192.168.0.10, 192.168.0.11, 192.168.0.12, 192.168.0.13]
```

In this option, how many trackingWorkers do the dataSources mean? four or two, or one?